compiler-core/src/type_.rs

```diff
@@ -46,25 +46,71 @@ pub trait HasType {
 
 #[derive(Debug, Clone, PartialEq)]
 pub enum Type {
+    /// App is the primary "wrapper" for most types. The type may come from a
+    /// third party module, but if `module` is empty that means the type is
+    /// coming from the gleam prelude instead (where Int, Float, String, etc is
+    /// defined). If the `App` takes generic arguments they will be put in the
+    /// `args` field.
+    ///
+    /// # Example
+    ///
+    /// ```gleam
+    /// Int //=> Type::App { name: "Int", args: vec![], module: vec![], public: true }
+    /// String //=> Type::App { name: "String", args: vec![], module: vec![], public: true }
+    /// ```
+    ///
     App {
         public: bool,
         module: Vec<String>,
         name: String,
         args: Vec<Arc<Type>>,
     },
 
+    /// Fn defines the type for an _anonymous_ lambda function. These should
+    /// not be confused with module functions which would instead be an
+    /// `Statement::Fn` with its own fields for its arguments and return type
+    ///
+    /// # Example
+    ///
+    /// ```gleam
+    /// fn(x: Int, y: Int) -> Int { x + y } //=> Type::Fn {
+    /// //   args: vec![
+    /// //     Type::App { name: "Int", args: vec![], module: vec![], public: true },
+    /// //     Type::App { name: "Int", args: vec![], module: vec![], public: true },
+    /// //   ],
+    /// //   retrn: Type::App { name: "Int", args: vec![], module: vec![], public: true },
+    /// // }
+    /// ```
+    ///
     Fn {
         args: Vec<Arc<Type>>,
         retrn: Arc<Type>,
     },
 
-    Var {
-        type_: Arc<RefCell<TypeVar>>,
-    },
+    /// A Var is a type variable, internally it holds a `TypeVar`, which could be
+    /// an `Unbound` variable (it is one specific type but we don't know what yet
+    /// in the inference process) a `Generic` variable (it stands in for any
+    /// possible type and cannot be specialised to any one type), or a `Link` (it
+    /// was an unbound variable but we worked out that it is some other type and
+    /// now we point to that one)
+    ///
+    Var { type_: Arc<RefCell<TypeVar>> },
```

Comment:
Better to keep these `TypeVar` docs with the `TypeVar`.
```suggestion
    /// A type variable. See the contained `TypeVar` enum for more information.
    ///
    Var { type_: Arc<RefCell<TypeVar>> },
```